CMakeLists.txt

```diff
@@ -2,6 +2,50 @@ cmake_minimum_required(VERSION 2.8.7)
 
 project(GSL CXX)
 
+# creates a library GSL which is an interface (header files only)
+add_library(GSL INTERFACE)
+
+# when minimum version required is 3.8.0 remove if below
+# both branches do exactly the same thing
+if ( CMAKE_MAJOR_VERSION VERSION_LESS 3.7.9)
+    if (NOT MSVC)
+        include(CheckCXXCompilerFlag)
+        CHECK_CXX_COMPILER_FLAG("-std=c++14" COMPILER_SUPPORTS_CXX14)
+        if(COMPILER_SUPPORTS_CXX14)
+            target_compile_options(GSL INTERFACE "-std=c++14")
+        else()
+            message(FATAL_ERROR "The compiler ${CMAKE_CXX_COMPILER} has no C++14 support. Please use a different C++ compiler.")
+        endif()
+        
+    endif()
+else ()
+    # set the GSL library to be compiled only with c++14
+    target_compile_features(GSL INTERFACE cxx_std_14)
+    # on *nix systems force the use of -std=c++XX instead of -std=gnu++XX (default)
+    set(CMAKE_CXX_EXTENSIONS OFF)
+endif()
+
+# add definitions to the library and targets that consume it
+target_compile_definitions(GSL INTERFACE
+    GSL_THROW_ON_CONTRACT_VIOLATION
```

Comment:
`GSL_THROW_ON_CONTRACT_VIOLATION` should be attached to `gsl_tests_config` instead of `GSL`.